helix-term/src/commands.rs

```diff
@@ -3284,33 +3286,88 @@ fn select_register(cx: &mut Context) {
     })
 }
 
-fn space_mode(cx: &mut Context) {
-    cx.on_next_key(move |cx, event| {
-        if let KeyEvent {
-            code: KeyCode::Char(ch),
-            ..
-        } = event
-        {
-            // TODO: temporarily show SPC in the mode list
-            match ch {
-                'f' => file_picker(cx),
-                'b' => buffer_picker(cx),
-                's' => symbol_picker(cx),
-                'w' => window_mode(cx),
-                'y' => yank_joined_to_clipboard(cx),
-                'Y' => yank_main_selection_to_clipboard(cx),
-                'p' => paste_clipboard_after(cx),
-                'P' => paste_clipboard_before(cx),
-                'R' => replace_selections_with_clipboard(cx),
-                // ' ' => toggle_alternate_buffer(cx),
-                // TODO: temporary since space mode took its old key
-                ' ' => keep_primary_selection(cx),
-                _ => (),
-            }
+macro_rules! mode_info {
+    // TODO: how to use one expr for both pat and expr?
+    // TODO: how to use replaced function name as str at compile time?
+    // TODO: extend to support multiple keys, but first solve the other two
+    {$name:literal, $cx:expr, $($key:expr => $func:expr; $funcs:literal),+,} => {
+        mode_info! {
+            $name, $cx,
+            $($key; $key => $func; $funcs,)+
         }
-    })
+    };
+    {$name:literal, $cx:expr, $($key:expr; $keyp:pat => $func:expr; $funcs:literal),+,} => {
+        $cx.editor.autoinfo = Some(Info::key(
+            $name,
+            vec![
+                $(
+                (vec![$key], $funcs),
+                )+
+            ],
+        ));
+        $cx.on_next_key(move |cx, event| {
+            match event {
+                $(
+                $keyp => $func(cx),
+                )+
+                _ => {}
+            }
+        })
+    }
 }
 
+fn space_mode(cx: &mut Context) {
+    mode_info! {
+        "space mode", cx,
+        key!('f'); key!('f') => file_picker; "file picker",
```

Comment:
Do we want the keys here to be hard-coded?  I realize the keymap config doesn't currently handle much outside of normal mode, so it might be a moot point.  But I wonder if there's a simple way to set ourselves up for success with that later.  (Maybe not.)

Reply:
Right now we don't have ways to configure this in the config file. But ideally this should be configurable. But that's after this patch I guess.

Reply:
Yeah, probably putting that off until we actually want to make that change makes more sense.